### PR TITLE
Clickable test names

### LINKED ISSUE #40: Clickable test names in pop-up which lead to the test (coverage)
# Description
For the pop-up, which appears during coverage visualisation in the gutter, create the ability to display tests that cover the line which you click the highlight on. The pop-up already displays the test names, these could be made clickable.

# Definition of Done
* [ ]  Tooltip which shows, which tests cover a line
* [ ]  Names of tests are pressable/clickable
* [ ]  Once clicked, the test name leads you to the actual test

### PR DESCRIPTION
**Pull request by bolkhod**
# Description of changes made
When you click on the green elements in the gutter (visualisation), a new popup menu appears. Test names which cover a line are now clickable. Clicking these test names makes the corresponding mini-editor in the toolwindow have a custom color (the same shade of green, for now). After 10 seconds, the color resets. This way a test code is highlighted for a short moment and can easily be spotted.

# Why is merge request needed
Because the client suggested this feature. This feature generally helps finding tests that cover a specific line (in the toolwindow).

Closes #40 

# Other notes
The UI of the popup menu has gotten uglier after the change.

# What is missing?
Custom colors can be added in the future - future option in settings
Better UI design for the popup menu

- [x] I have checked that I am merging into correct branch

### CODE REVIEW

**Comment by martinmladenov** [comment]
Is this PR ready for review?

**Comment by bolkhod** [comment]
> Is this PR ready for review?

The core of the code is. I am just aware that some other PRs (which will likely be merged prior to this one) will have conflicts.
I aim to merge dev into the branch and solve them, before fully opening.
But you can check the added stuff already, its shouldn't change (much)

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/listener/TestGenerationResultListenerImpl.kt [hunk 1]
```diff
@@ -17,9 +17,8 @@ class TestGenerationResultListenerImpl(private val project: Project) : TestGener
 
         val testCaseDisplayService = project.service<TestCaseDisplayService>()
 
-        val tests = testReport.testCaseList.values.map { it.testCode }
         ApplicationManager.getApplication().invokeLater {
-            testCaseDisplayService.displayTestCases(tests)
+            testCaseDisplayService.displayTestCases(testReport)
```
Comment (jzelenjak) [inline src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/listener/TestGenerationResultListenerImpl.kt:21]: This is much better to read!

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/TestCaseDisplayService.kt [hunk 2]
```diff
@@ -54,4 +61,25 @@ class TestCaseDisplayService(private val project: Project) {
...
+     * @param name name of the test whose editor should be highlighted
+     */
+    fun highlight(name: String) {
+        for (i in editorList) {
+            val testCase = i.first
+            if (testCase == name) {
+                val editor = i.second
+                val backgroundDefault = editor.background
+                editor.background = Color(100, 150, 20, 30)
+                Thread {
```
Comment (jzelenjak) [inline src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/TestCaseDisplayService.kt:77]: I like this way of doing it. Well done!

**Review by jzelenjak — APPROVED** [review]
The colouring of the selected test looks very cool in my opinion.  Especially, implementing the reset in a thread was a very clever approach.
Good job!

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/TestCaseDisplayService.kt (hunk 2 above)
Comment (bolkhod) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/TestCaseDisplayService.kt:77]: thanks

**Review by jzelenjak — CHANGES_REQUESTED** [review]
After running your code, i noticed that there is an issue that it is not possible to highlight the test when running the generation more than once. This has to be addressed.

Otherwise, everything looks good to me

**Review by jzelenjak — DISMISSED** [review]
Thank you for fixing the bug!

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/TestCaseDisplayService.kt [hunk 3]
```diff
@@ -54,4 +62,25 @@ class TestCaseDisplayService(private val project: Project) {
...
+     *
+     * @param name name of the test whose editor should be highlighted
+     */
+    fun highlight(name: String) {
+        for (i in editorList) {
+            val testCase = i.first
+            if (testCase == name) {
+                val editor = i.second
+                val backgroundDefault = editor.background
+                editor.background = Color(100, 150, 20, 30)
```
[the remark this answers is not in the document]
Comment (bolkhod) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/TestCaseDisplayService.kt]: great idea! I have done exactly that

File: src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/coverage/CoverageRenderer.kt [hunk 4]
```diff
@@ -34,8 +38,17 @@ class CoverageRenderer(private val color: Color, private val lineNumber: Int, pr
      */
     override fun doAction(editor: Editor, e: MouseEvent) {
         e.consume()
-        val panel = FormBuilder.createFormBuilder().addComponent(JBLabel(" Covered by tests: $tests "), 10)
-            .addVerticalGap(10).panel
+        val prePanel = FormBuilder
+                .createFormBuilder()
+                .addComponent(JBLabel(" Covered by tests:"), 10)
+
+        for (i in tests) {
```
[the remark this answers is not in the document]
Comment (bolkhod) [reply src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/coverage/CoverageRenderer.kt]: yes, this is now `testNames` to make it clear to other devs

**Review by kirilvasilev16 — APPROVED** [review]
Thank you for incorporating my feedback!

**Review by jzelenjak — APPROVED** [review]

**Pull request merged**